core/metacling/src/TClingCallFunc.cxx

```diff
@@ -1210,7 +1149,7 @@ tcling_callfunc_ctor_Wrapper_t TClingCallFunc::make_ctor_wrapper(const TClingCla
    if (const TypeDecl *TD = dyn_cast<TypeDecl>(info->GetDecl())) {
       // This is a class, struct, or union member.
       QualType QT(TD->getTypeForDecl(), 0);
-      ROOT::TMetaUtils::GetNormalizedName(class_name, QT, *fInterp, fNormCtxt);
```

Comment:
Some of the types must not be spelled out with all their template arguments - which is registered in `fNormCtxt`. ATLAS has such a case. @pcanal do you remember the details?

Reply:
`fNormCtxt` tells ROOT to do two things: a) partially desugar the types; b) qualify the types so that the wrapper compiles. We do not practially care about a) (if the wrapper will contain `Double32_t` or `double` and the full type qualification happens in `GetTypeAsString` which calls `cling::utils::Transform::GetPartiallyDesugaredType`.

In fact `hist/histdrawv7/test/histhistdrawv7testUnit` has a test which catches such issues - it just uses horribly long typenames along with inheritance via template parameter packs...

Reply:
The issue was:

    In CallFunc remove default arg for __shared_ptr.
    
    We need to manipulate the fully qualified name to remove the default template
    to properly handle the __shared_ptr case until the clang type printer properly
    handles template parameter that are enumerator.

Reply:
@pcanal did you mean 291f23eca4b54ea5166c1f97f01a950620032375

Reply:
The 'real' commit is 'a211a41386e46f8467fdd0c2230d4432d521f542'.   https://github.com/root-project/root/commit/291f23eca4b54ea5166c1f97f01a950620032375 is an improvement on the first one (probably because `GetNormalizedName` is more accurate/better tested than `R__DropDefaultArg`.

The question is whether Clang has improved enough that now, the code generated for the function/method of `__shared_ptr` (or maybe any template) with a (default) template parameter that are enumeraton will now compiled.

Reply:
How we can check that?

Reply:
I assume that using some CallFunc on `shared_ptr<enumtype>` or `shared_ptr<std::vector<enumtype>>` and their base class should be 'enough'.  (technically, but impractically, running the same test with a211a41386e46f8467fdd0c2230d4432d521f542^ and seeing it fail).

Reply:
I could not build that commit on my platform unfortunately. I added a test which I believe captures the use-case and produces a wrapper which compiles fine:
```cpp
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wformat-security"
__attribute__((used)) __attribute__((annotate("__cling__ptrcheck(off)")))
extern "C" void __cf_11(void* obj, int nargs, void** args, void* ret)
{
   ((void (&)(std::shared_ptr<std::vector<E, std::allocator<E> > >))add_sp)(*(std::shared_ptr<std::vector<E, std::allocator<E> > >*)args[0]);
   return;
}
#pragma clang diagnostic pop
```

Reply:
Fair enough, it looks (so far) that it cover the case referred to.